### PR TITLE
Add CAN bus.

### PR DESCRIPTION
**Pull request by matoushybl**
Adds CAN bus as I developed it for my project.

Any feedback is welcome. I'll continue the development to make it fully working in context of the HAL.

Also CAN traits are under review in embedded-hal, so I might implement the traits before they are merged for future compatibility.

### CODE REVIEW

File: src/i2c_slave.rs [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+        self.state = None;
+        self.buffer_index = 0;
+        self.transfer_len = 0;
+        data
+    }
+
+    pub fn get_state(&self) -> Option<State> {
+        self.state
+    }
+}
```
Comment (jonas-schievink) [inline src/i2c_slave.rs:383]: Unrelated change?

File: src/lib.rs [hunk 2]
```diff
@@ -84,6 +86,12 @@ pub mod usb;
 #[cfg(feature = "device-selected")]
 pub mod watchdog;
 
+// TODO other controllers have CAN
+#[cfg(any(
+    feature = "stm32f091",
```
Comment (jonas-schievink) [inline src/lib.rs:91]:
```suggestion
    feature = "stm32f091",
    feature = "stm32f042",
    feature = "stm32f072",
```

That's all of them according to the ST website.

**Comment by matoushybl** [comment]
Hi, @jonas-schievink! Thanks for the review! I have fixed the problems you've found and generally fixed some problems with the code. It is still not mergeable to master (pin definitions are still not implemented), but I expect to finish it sometime next week.

**Review by therealprof — CHANGES_REQUESTED** [review]
Missing a CHANGELOG entry. ;)

**Review by therealprof — APPROVED** [review]
Haven't had a chance to test it out in person but let's get this in so people can play with it.

Thanks a lot!

bors r+

**Pull request merged**